app/Rules/UsernameFormatRule.php

```diff
@@ -5,13 +5,56 @@
 use App\Actions\GenerateUsername;
 use Closure;
 use Illuminate\Contracts\Validation\ValidationRule;
+use Illuminate\Contracts\Validation\ValidatorAwareRule;
+use Illuminate\Support\Facades\Validator as FacadeValidator;
+use Illuminate\Validation\Validator;
 
-class UsernameFormatRule implements ValidationRule
+class UsernameFormatRule implements ValidationRule, ValidatorAwareRule
 {
+    protected Validator $validator;
+
+    protected array $rules = [
+        'required',
+        'string',
+        'min:2',
+        'max:50',
+        'unique:users,username',
+    ];
+
     public function validate(string $attribute, mixed $value, Closure $fail): void
     {
-        if (! is_string($value) || $value !== (new GenerateUsername())($value)) {
-            $fail('The :attribute must be valid');
+        $validator = FacadeValidator::make(
+            [$attribute => $value],
+            [$attribute => $this->rules],
+        )->after(function () use ($value, $fail) {
+            if (! is_string($value)) {
+                return;
+            }
+
+            if ($value !== (new GenerateUsername())($value)) {
+                $fail(
+                    __(
+                        'The :attribute must be valid. Only lowercase ASCII characters are allowed. Hyphens can be used. Whitespace, underscores, and multiple hyphens are not permitted. The username cannot start or end with a hyphen.'
+                    )
+                );
+            }
+        });
+
+        if ($validator->fails()) {
+            $this->validator->errors()->merge($validator->errors());
```

Comment:
I guess the reason for making this rule `ValidatorAware` is so that the error can be added to the global validator, which is used by Livewire?

Reply:
Oh this is ok :) I was just curious ;)